File: packages/python/elizaos/features/advanced_capabilities/actions/update_entity.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING
from uuid import UUID

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import UPDATE_ENTITY_TEMPLATE
from elizaos.types import Action, ActionExample, ActionResult, Content, ModelType
from elizaos.utils.spec_examples import convert_spec_examples
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_action_spec("UPDATE_ENTITY")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    return convert_spec_examples(_spec)


@dataclass
class UpdateEntityAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        return message.entity_id is not None

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        if state is None:
            raise ValueError("State is required for UPDATE_ENTITY action")

        entity_id = message.entity_id
        if not entity_id:
            return ActionResult(
                text="No entity specified to update",
                values={"success": False, "error": "no_entity_id"},
                data={"actionName": "UPDATE_ENTITY"},
                success=False,
            )

        entity = await runtime.get_entity(entity_id)
        if entity is None:
            return ActionResult(
                text="Entity not found",
                values={"success": False, "error": "entity_not_found"},
                data={"actionName": "UPDATE_ENTITY"},
                success=False,
            )

        state = await runtime.compose_state(
            message, ["RECENT_MESSAGES", "ACTION_STATE", "ENTITY_INFO"]
        )

        entity_info = f"""
Entity ID: {entity.id}
Name: {entity.name or "Unknown"}
Type: {entity.entity_type or "Unknown"}
"""
        if entity.metadata:
            entity_info += f"Metadata: {entity.metadata}"

        template = (
            runtime.character.templates.get("updateEntityTemplate")
            if runtime.character.templates and "updateEntityTemplate" in runtime.character.templates
            else UPDATE_ENTITY_TEMPLATE
        )
        prompt = runtime.compose_prompt(state=state, template=template)
        prompt = prompt.replace("{{entityInfo}}", entity_info)

        response_text = await runtime.use_model(ModelType.TEXT_LARGE, prompt=prompt)
        parsed_xml = parse_key_value_xml(response_text)

        if parsed_xml is None:
            raise ValueError("Failed to parse XML response")

        thought = str(parsed_xml.get("thought", ""))
        target_entity_id_str = str(parsed_xml.get("entity_id", str(entity_id)))

        target_entity_id = UUID(target_entity_id_str)

        updates_raw: object = parsed_xml.get("updates", {})
        updated_fields: list[str] = []

        if isinstance(updates_raw, dict):
            field_list = updates_raw.get("field", [])
            if isinstance(field_list, dict):
                field_list = [field_list]
            for field_update in field_list:
                if isinstance(field_update, dict):
                    field_name = str(field_update.get("name", ""))
                    field_value = str(field_update.get("value", ""))
                    if field_name and field_value:
                        if entity.metadata is None:
                            entity.metadata = {}
                        entity.metadata[field_name] = field_value
                        updated_fields.append(field_name)

        if not updated_fields:
            return ActionResult(
                text="No fields to update",
                values={"success": True, "noChanges": True},
                data={"actionName": "UPDATE_ENTITY", "thought": thought},
                success=True,
            )

        await runtime.update_entity(entity)

        response_content = Content(
            text=f"Updated entity fields: {', '.join(updated_fields)}",
            actions=["UPDATE_ENTITY"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Updated entity: {', '.join(updated_fields)}",
            values={
                "success": True,
                "entityUpdated": True,
                "entityId": str(target_entity_id),
                "updatedFields": ", ".join(updated_fields),
            },
            data={
                "actionName": "UPDATE_ENTITY",
                "entityId": str(target_entity_id),
                "updatedFields": updated_fields,
                "thought": thought,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


update_entity_action = Action(
    name=UpdateEntityAction.name,
    similes=UpdateEntityAction().similes,
    description=UpdateEntityAction.description,
    validate=UpdateEntityAction().validate,
    handler=UpdateEntityAction().handler,
    examples=UpdateEntityAction().examples,
)

File: packages/rust/src/features/advanced_capabilities/actions/update_entity.rs
//! UPDATE_ENTITY action implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;
use std::sync::Arc;
use uuid::Uuid;

use crate::error::{PluginError, PluginResult};
use crate::generated::spec_helpers::require_action_spec;
use crate::prompts::UPDATE_ENTITY_TEMPLATE;
use crate::runtime::{IAgentRuntime, ModelParams};
use crate::types::{ActionResult, Memory, ModelType, State};
use crate::xml::parse_key_value_xml;

use super::Action;

/// Action for updating entity information.
pub struct UpdateEntityAction;

static SPEC: Lazy<&'static crate::generated::spec_helpers::ActionDoc> =
    Lazy::new(|| require_action_spec("UPDATE_ENTITY"));

#[async_trait]
impl Action for UpdateEntityAction {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn similes(&self) -> &[&'static str] {
        static SIMILES: Lazy<Box<[&'static str]>> = Lazy::new(|| {
            SPEC.similes
                .iter()
                .map(|s| s.as_str())
                .collect::<Vec<_>>()
                .into_boxed_slice()
        });
        &SIMILES
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    async fn validate(&self, _runtime: &dyn IAgentRuntime, message: &Memory) -> bool {
        message.entity_id.is_some()
    }

    async fn handler(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        message: &Memory,
        state: Option<&State>,
        _responses: Option<&[Memory]>,
    ) -> PluginResult<ActionResult> {
        let _state = state.ok_or_else(|| {
            PluginError::StateRequired("State is required for UPDATE_ENTITY action".to_string())
        })?;

        let entity_id = message.entity_id.ok_or_else(|| {
            PluginError::InvalidInput("No entity specified to update".to_string())
        })?;

        let entity = runtime
            .get_entity(entity_id)
            .await?
            .ok_or_else(|| PluginError::NotFound("Entity not found".to_string()))?;

        // Compose state
        let composed_state = runtime
            .compose_state(message, &["RECENT_MESSAGES", "ACTION_STATE", "ENTITY_INFO"])
            .await?;

        // Build entity info for prompt
        let entity_info = format!(
            "Entity ID: {}\nName: {}\nType: {}",
            entity.id,
            entity.name.as_deref().unwrap_or("Unknown"),
            entity.entity_type.as_deref().unwrap_or("Unknown")
        );

        // Get template and compose prompt
        let template = runtime
            .character()
            .templates
            .get("updateEntityTemplate")
            .map(|s| s.as_str())
            .unwrap_or(UPDATE_ENTITY_TEMPLATE);

        let prompt = runtime
            .compose_prompt(&composed_state, template)
            .replace("{{entityInfo}}", &entity_info);

        // Call the model
        let response = runtime
            .use_model(ModelType::TextLarge, ModelParams::with_prompt(&prompt))
            .await
            .map_err(|e| PluginError::ModelError(e.to_string()))?;

        let response_text = response
            .as_text()
            .ok_or_else(|| PluginError::ModelError("Expected text response".to_string()))?;

        // Parse XML response
        let parsed = parse_key_value_xml(response_text)
            .ok_or_else(|| PluginError::XmlParse("Failed to parse response XML".to_string()))?;

        let thought = parsed.get("thought").cloned().unwrap_or_default();
        let target_entity_id_str = parsed
            .get("entity_id")
            .cloned()
            .unwrap_or_else(|| entity_id.to_string());
        let field_name = parsed.get("field_name").cloned();
        let field_value = parsed.get("field_value").cloned();

        // Validate target entity ID
        let target_entity_id = Uuid::parse_str(&target_entity_id_str).map_err(|_| {
            PluginError::InvalidInput(format!("Invalid entity ID: {}", target_entity_id_str))
        })?;

        if let (Some(name), Some(value)) = (field_name.clone(), field_value.clone()) {
            // Update entity metadata
            let mut updated_entity = entity.clone();
            updated_entity
                .metadata
                .insert(name.clone(), serde_json::json!(value));
            runtime.update_entity(&updated_entity).await?;

            Ok(
                ActionResult::success(format!("Updated entity field: {}", name))
                    .with_value("success", true)
                    .with_value("entityUpdated", true)
                    .with_value("entityId", target_entity_id.to_string())
                    .with_value("updatedField", name.clone())
                    .with_data("actionName", "UPDATE_ENTITY")
                    .with_data("entityId", target_entity_id.to_string())
                    .with_data("fieldName", name)
                    .with_data("fieldValue", value)
                    .with_data("thought", thought),
            )
        } else {
            Ok(ActionResult::success("No fields to update")
                .with_value("success", true)
                .with_value("noChanges", true)
                .with_data("actionName", "UPDATE_ENTITY")
                .with_data("thought", thought))
        }
    }
}
